### ISSUE TITLE
twisted.cred._digest miscorrectly handles md5-sess

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7835]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7835|
|Type|defect|
|Created|2015-03-26 09:21:20Z|
|Branch|https://github.com/twisted/twisted/tree/tcdigest-md5sess-7835|

The md5-sess algorithm has HA1=MD5(MD5(username:realm:password):nonce:cnonce). Twisted follows the RFC's example code which has an errata filed (http://www.rfc-editor.org/errata_search.php?rfc=2617&eid=1649). The RFC says that the representation for all MD5 outputs is hex, but Twisted does it as raw bytes.

<details><summary>Searchable metadata</summary>

```
trac-id__7835 7835
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_tcdigest_md5sess_7835 branches/tcdigest-md5sess-7835
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1427361680973783 1427361680973783
changetime__1427693728929484 1427693728929484
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44287]) Branching to tcdigest-md5sess-7835.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Fix done, builders spun, please review...un. (Couldn't get it all to rhyme :( )

This work sponsored by Tavendo GmbH.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(Also, this should satisfy https://twistedmatrix.com/trac/wiki/CompatibilityPolicy#FixingGrossViolationofSpecifications)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|


Thanks for working on this!

------

```
twisted.cred.credentials.DigestedCredentials incorrectly handled md5-sess hashing according to the RFC, which has now been fixed.

```

maybe

```
twisted.cred.credentials.DigestedCredentials now handles md5-sess hashing according to the RFC 2617 erata. Previous implementation was based on the RFC example.
```

--------

To have a valid review for this I would like to construct an example which I can use to text the code on trunk and on this branch with other implementations (browsers/curl)

It would have been great if the review request would have also included an example to run a manual functional test.

I will look at creating one and maybe submit it at https://twistedmatrix.com/documents/current/web/examples/index.html

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

Hi Adi,

I don't think we really need to mention the errata.

An external test might be good, but I'm not sure how it is actually usable -- rarely ever touch cred.

-hawkie

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Ok. Here is my short example https://gist.github.com/adiroiban/9ee8e29af3193a258f68 

based on http://twistedmatrix.com/documents/current/web/howto/web-in-60/http-auth.html

I can confirm it fails on trunk and pass on your branch.

Looks good. Please merge. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44321]) Merging tcdigest-md5sess-7835: twisted.cred._digest miscorrectly handles md5-sess

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7835](https://github.com/twisted/twisted/issues/7835)
